applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_lagrange_dirichlet_condition.cpp

```diff
@@ -0,0 +1,458 @@
+        for ( unsigned int i = 0; i < number_of_nodes; i++ )
+        {
+            if (Variables.N[i] > std::numeric_limits<double>::epsilon() )
+            {
+                for ( unsigned int j = 0; j < dimension; j++)
+                {
+                    field_displacement[j] += Variables.N[i] * Variables.CurrentDisp(i,j);
+                }
+            }
+        }
+
+        const double penetration = MathUtils<double>::Dot((field_displacement - imposed_displacement), unit_normal_vector);
+
+        // If penetrates, apply constraint, otherwise no
+        if (penetration >= 0.0)
+        {
+            apply_constraints = false;
+        }
+    }
+
+    if (apply_constraints)
+    {
+        Matrix lagrange_matrix = ZeroMatrix(matrix_size, matrix_size);
+
+         // Loop over Lagrange Multipliers
+        for (unsigned int i = 0; i < number_of_nodes; i++)
+        {
+            // Loop over shape functions of displacements
+            for (unsigned int j = 0; j < number_of_nodes; j++)
+            {
+                const double NN = Variables.N[j] * Variables.N[i];
+                const unsigned int ibase = i * dimension + dimension * number_of_nodes;
+                const unsigned int jbase = j * dimension;
+
+                // Matrix in following shape:
+                // |aug*H   H^T|
+                // |H       0  |
+                // Lambda in X
+                for (unsigned int k = 0; k < dimension; k++)
+                {
+                    lagrange_matrix(jbase+k, jbase+k) = NN * augmentation_factor;
+                    lagrange_matrix(ibase+k, jbase+k) = NN;
+                    lagrange_matrix(jbase+k, ibase+k) = NN;
+                }
+
+
+                //Stabilization needed for some solvers
+                if (stabilization)
+                {
+                    lagrange_matrix(ibase, ibase) = -std::numeric_limits<double>::epsilon();
+                    lagrange_matrix(ibase + 1, ibase + 1) = -std::numeric_limits<double>::epsilon();
+
+                    if (dimension == 3)
+                            lagrange_matrix(ibase +2, ibase +2) = -std::numeric_limits<double>::epsilon();
+
+                }
+            }
+        }
+
+        lagrange_matrix  *= this->GetIntegrationWeight();
+
+        // Calculate LHS Matrix and RHS Vector
+        if ( CalculateStiffnessMatrixFlag == true )
+        {
+            rLeftHandSideMatrix = lagrange_matrix;
+        }
+
+        if ( CalculateResidualVectorFlag == true )
+        {
+            Vector gap_function = ZeroVector(matrix_size);
+            for (unsigned int i = 0; i < number_of_nodes; i++)
+            {
+                const array_1d<double, 3> disp = GetGeometry()[i].FastGetSolutionStepValue(DISPLACEMENT);
+                const int index = dimension * i;
+
+                const array_1d<double, 3> lagrange_multiplier = GetGeometry()[i].FastGetSolutionStepValue(VECTOR_LAGRANGE_MULTIPLIER);
```

Comment:
```suggestion
                const array_1d<double, 3>& lagrange_multiplier = GetGeometry()[i].FastGetSolutionStepValue(VECTOR_LAGRANGE_MULTIPLIER);
```

Something bigger than a double should have a reference